src/k8s/pkg/k8sd/features/calico/network.go

```diff
@@ -10,28 +10,53 @@ import (
 	"github.com/canonical/k8s/pkg/utils"
 )
 
+const (
+	enabledMsg          = "enabled"
+	disabledMsg         = "disabled"
+	deployFailedMsgTmpl = "Failed to deploy Calico, the error was: %v"
+	deleteFailedMsgTmpl = "Failed to delete Calico, the error was: %v"
+)
+
 // ApplyNetwork will deploy Calico when cfg.Enabled is true.
 // ApplyNetwork will remove Calico when cfg.Enabled is false.
-// ApplyNetwork returns an error if anything fails.
-func ApplyNetwork(ctx context.Context, snap snap.Snap, cfg types.Network, annotations types.Annotations) error {
+// ApplyNetwork will always return a FeatureStatus indicating the current status of the
+// deployment.
+// ApplyNetwork returns an error if anything fails. The error is also wrapped in the .Message field of the
+// returned FeatureStatus.
+func ApplyNetwork(ctx context.Context, snap snap.Snap, cfg types.Network, annotations types.Annotations) (types.FeatureStatus, error) {
+	status := types.FeatureStatus{
+		Version: calicoTag,
+		Enabled: cfg.GetEnabled(),
+	}
+
 	m := snap.HelmClient()
 
 	if !cfg.GetEnabled() {
 		if _, err := m.Apply(ctx, chartCalico, helm.StateDeleted, nil); err != nil {
-			return fmt.Errorf("failed to uninstall network: %w", err)
+			applyErr := fmt.Errorf("failed to uninstall network: %w", err)
+			status.Message = fmt.Sprintf(deleteFailedMsgTmpl, applyErr)
```

Comment:
much better to inline the messages here. when you then have an error message, it's easier to grep the code. also, avoid reusing the same error message in multiple places. e.g. this could be

```
return FeatureStatus{Message: fmt.Sprintf("failed to delete Calico: %v", err)}, err
```